### PR TITLE
fix(addon-editor): prevent re-attach elements after state changes

### PR DESCRIPTION
**Pull request by splincode**
## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] The commit message follows [Conventional Commits](https://www.conventionalcommits.org/en/)
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type

What kind of change does this PR introduce?

- [x] Bugfix
- [ ] Feature
- [ ] Refactoring
- [ ] Code style update
- [ ] Build or CI related changes
- [ ] Documentation content changes

## What is the current behavior?

Part of #1646

https://user-images.githubusercontent.com/12021443/169816885-cc16a01f-0e6d-49ec-aa15-6b51c5d5b286.mov

## What is the new behavior?

https://user-images.githubusercontent.com/12021443/169817035-297faee5-f6c4-44ab-8eeb-ead1324fa2bd.mov

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: projects/addon-editor/components/editor-socket/editor-socket.component.ts [hunk 1]
```diff
@@ -25,18 +25,14 @@ import {TUI_SANITIZER} from '@taiga-ui/core';
     },
 })
 export class TuiEditorSocketComponent {
+    private html = '';
+
     @Input()
     set content(content: string) {
```
Comment (waterplea) [inline projects/addon-editor/components/editor-socket/editor-socket.component.ts:31]: Let's make it declarative, remove setter from `content`, add getter for `html` with `@HostBinding('innerHTML')` and make a private method with `@tuiPure` to sanitize content.

File: projects/demo/src/modules/components/editor-new/examples/1/index.html [hunk 2]
```diff
@@ -13,6 +13,7 @@
     exampleText="exampleText"
     class="editor"
     [formControl]="control"
+    (ngModelChange)="updateView()"
```
Comment (waterplea) [inline projects/demo/src/modules/components/editor-new/examples/1/index.html:16]: You cannot mix reactive and template forms

File: projects/addon-editor/components/editor-new/editor-new.component.ts [hunk 3]
```diff
@@ -116,6 +116,17 @@ export class TuiEditorNewComponent
         this.updateFocused(active);
     }
 
+    /**
+     * @note:
+     * Editor invalidates html attributes inside
+     * and overwrites again html content
+     */
+    onStateChange(): void {
+        if (!!this.editor?.html && this.editor.html !== this.value) {
```
Comment (waterplea) [inline projects/addon-editor/components/editor-new/editor-new.component.ts:125]: Is there really a case when `stateChange` causes content change?

File: projects/demo/src/modules/components/editor-new/examples/1/index.ts [hunk 4]
```diff
@@ -32,9 +32,15 @@ export class TuiEditorNewExample1 {
...
+    constructor(@Inject(ChangeDetectorRef) private readonly cd: ChangeDetectorRef) {}
+
     onClick(): void {
         this.control.setValue(
             '<p>Control</p><h2>is</h2><h1><span style="color: #e01f19">updated</span></h1>',
         );
     }
+
+    updateView(): void {
+        this.cd.detectChanges();
```
Comment (waterplea) [inline projects/demo/src/modules/components/editor-new/examples/1/index.ts:44]: Why is this required?

File: projects/demo/src/modules/components/editor-new/examples/2/index.html [hunk 5]
```diff
@@ -3,6 +3,7 @@
     class="editor"
     [formControl]="control"
     [tools]="builtInTools"
+    (ngModelChange)="updateView()"
```
Comment (waterplea) [inline projects/demo/src/modules/components/editor-new/examples/2/index.html:6]: Same here

File: projects/demo/src/modules/components/editor-new/examples/2/index.ts [hunk 6]
```diff
@@ -24,4 +24,10 @@ export async function importStarterKit(): Promise<unknown> {
 export class TuiEditorNewExample2 {
     readonly builtInTools = [TuiEditorTool.Undo];
     readonly control = new FormControl('', Validators.required);
+
+    constructor(@Inject(ChangeDetectorRef) private readonly cd: ChangeDetectorRef) {}
+
+    updateView(): void {
+        this.cd.detectChanges();
```
Comment (waterplea) [inline projects/demo/src/modules/components/editor-new/examples/2/index.ts:31]: Same question

File: projects/demo/src/modules/components/editor-new/examples/3/index.html [hunk 7]
```diff
@@ -3,6 +3,7 @@
     class="editor"
     [formControl]="control"
     [tools]="builtInTools"
+    (ngModelChange)="updateView()"
```
Comment (waterplea) [inline projects/demo/src/modules/components/editor-new/examples/3/index.html:6]: And here

File: projects/demo/src/modules/components/editor-new/examples/3/index.ts [hunk 8]
```diff
@@ -62,4 +63,8 @@ export class TuiEditorNewExample3 {
             `);
         });
     }
+
+    updateView(): void {
+        this.cd.detectChanges();
```
Comment (waterplea) [inline projects/demo/src/modules/components/editor-new/examples/3/index.ts:68]: And here :)

**Review by waterplea — CHANGES_REQUESTED** [review]

File: projects/addon-editor/components/editor-new/editor-new.component.ts (hunk 3 above)
Comment (splincode) [reply projects/addon-editor/components/editor-new/editor-new.component.ts:125]: Yes, e2e tests therefore fell

File: projects/addon-editor/components/editor-socket/editor-socket.component.ts (hunk 1 above)
Comment (splincode) [reply projects/addon-editor/components/editor-socket/editor-socket.component.ts:46]: unfortunately `@HostBinding('innerHTML')`
doesn't cause a repaint view, if the change detection
wave wasn't triggered

File: projects/addon-editor/components/editor-socket/editor-socket.component.ts [hunk 9]
```diff
@@ -23,20 +25,24 @@ import {TUI_SANITIZER} from '@taiga-ui/core';
     host: {
         class: 'tui-editor-socket',
     },
+    providers: [TuiDestroyService],
 })
 export class TuiEditorSocketComponent {
     @Input()
+    set editor(editor: AbstractTuiControl<string>) {
+        editor.control?.valueChanges
+            .pipe(takeUntil(this.destroy$))
```
Comment (nsbarsukov) [inline projects/addon-editor/components/editor-socket/editor-socket.component.ts:34]: `this.destroy$` will kill subscription on destroy of the component. It is nice.

But what happens if Input-property changes twice (or more) during the component's life?
Will the component have only single subscription ?

File: projects/addon-editor/components/editor-new/editor-new.component.html [hunk 10]
```diff
@@ -51,8 +51,8 @@
                     class="tui-editor-socket"
                     [value]="value"
                     [editable]="interactive"
-                    (valueChange)="onModelChange($event)"
-                    (stateChange)="(0)"
+                    (valueChange)="(0)"
```
Comment (waterplea) [inline projects/addon-editor/components/editor-new/editor-new.component.html:54]: Are we sure we don't care about `valueChange`? 🤔

File: projects/addon-editor/components/editor-new/editor-new.component.ts [hunk 11]
```diff
@@ -116,8 +116,15 @@ export class TuiEditorNewComponent
...
     }
 
+    /**
+     * @note:
+     * TipTap editor can invalidate html attributes inside
+     * provided content, in order for the ViewModel and Model to be synchronized
+     */
     onModelChange(value: string): void {
-        this.updateValue(value);
+        if (value !== this.value) {
```
Comment (waterplea) [inline projects/addon-editor/components/editor-new/editor-new.component.ts:125]: `updateValue` from `AbstractTuiControl` already performs that check

File: projects/addon-editor/components/editor-socket/editor-socket.component.ts (hunk 1 above)
Comment (waterplea) [reply projects/addon-editor/components/editor-socket/editor-socket.component.ts:44]: But how can `@Input` change without change detection?

File: projects/addon-editor/components/editor-socket/editor-socket.component.ts [hunk 12]
```diff
@@ -23,20 +25,24 @@ import {TUI_SANITIZER} from '@taiga-ui/core';
     host: {
         class: 'tui-editor-socket',
     },
+    providers: [TuiDestroyService],
 })
 export class TuiEditorSocketComponent {
     @Input()
+    set editor(editor: AbstractTuiControl<string>) {
```
Comment (waterplea) [inline projects/addon-editor/components/editor-socket/editor-socket.component.ts:32]: This is wrong, socket is for displaying content that was created by editor, there might be no editor at all when it's used.

**Review by waterplea — CHANGES_REQUESTED** [review]

File: projects/addon-editor/components/editor-socket/editor-socket.component.ts (hunk 1 above)
Comment (splincode) [reply projects/addon-editor/components/editor-socket/editor-socket.component.ts:44]: very strange, I have warning after sanitizing when use `@HostBinding('innerHTML')` 
`WARNING: sanitizing HTML stripped some content, see http://g.co/ng/security#xss`
